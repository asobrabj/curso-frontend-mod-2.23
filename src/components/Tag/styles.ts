import styled from 'styled-components'

interface ITagStyle {
  $background: string
}

export const Tag = styled.span<ITagStyle>`
  text-transform: capitalize;
  padding: 2px 9px;
  background-color: ${(props) => props.theme.color[props.$background]};
  border: 1px solid ${(props) => props.theme.color.shadow};
  border-radius: 10px;
  font-weight: bold;
  margin-right: 8px;
  cursor: pointer;
`
