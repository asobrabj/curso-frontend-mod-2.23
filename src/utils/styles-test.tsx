import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
