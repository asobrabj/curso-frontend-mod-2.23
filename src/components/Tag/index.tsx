import { useStorage } from '../../context/localStorage/localStorage'
import * as S from './styles'

interface ITag {
  name: string
  text: string
  background: string
}

export default function Tag({ name, text, background }: ITag) {
  const { getTags } = useStorage()
  const translate = {
    categoria: 'category',
    prioridade: 'priority',
  }

  return (
    <S.Tag
      $background={background}
      onClick={() => getTags(translate[name], text)}
    >
      {text}
    </S.Tag>
  )
}
