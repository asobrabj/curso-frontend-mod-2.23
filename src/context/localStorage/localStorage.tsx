import { createContext, useContext, useEffect, useState } from 'react'
import { IFormData } from '../../components/FormAddAnnotation'
import IList from '../../types/lists'

export interface ILocalStorageContext {
  anotations: Array<IFormData> | null
  setAnotations: (data: IFormData | Array<IFormData>) => void
  getFilters: (value: string) => void
  getTags: (key: 'category' | 'priority', value: string) => void
  getStorage: (key: string) => Array<IFormData | IList> | string | null
  setStorage: (key: string, data: Array<IFormData | IList> | string) => void
}

interface ILocalStorageProvider {
  children: React.ReactNode
}

const LocalStorageContext = createContext<ILocalStorageContext>(
  {} as ILocalStorageContext
)

export function LocalStorageProvider({ children }: ILocalStorageProvider) {
  const [anotations, setAnotations] = useState<Array<IFormData> | null>([])

  useEffect(() => {
    const anotations = getStorage('anotations')
    setAnotations(anotations)
  }, [])

  const getTags = (key: 'category' | 'priority', value: string) => {
    const anotations = getStorage('anotations')
    const data = anotations?.filter(
      (item: IFormData) =>
        item[key]?.toLowerCase().includes(value.toLowerCase())
    )

    if (data) setAnotations(data)
    return
  }

  const getFilters = (value: string) => {
    const anotations = getStorage('anotations')
    const name = anotations?.filter((item: IFormData) =>
      item.name.toLowerCase().includes(value.toLowerCase())
    )

    if (name) setAnotations(name)
    return
  }

  const setAnotationsFn = (data: IFormData | Array<IFormData>) => {
    const storage = getStorage('anotations')
    let newStorage: Array<IFormData> = []
    if (storage) {
      newStorage = [...storage, data]
    }

    if (data instanceof Array) {
      newStorage = data
    }

    setAnotations(newStorage)
    localStorage.setItem('anotations', JSON.stringify(newStorage))
  }

  const getStorage = (key: string) => {
    const storedData = localStorage.getItem(key)
    if (storedData) {
      return JSON.parse(storedData)
    }
    return []
  }

  const saveToStorage = (
    key: string,
    data: Array<IFormData | IList> | string
  ) => {
    localStorage.setItem(key, JSON.stringify(data))
    return
  }
  return (
    <LocalStorageContext.Provider
      value={{
        getTags,
        getFilters,
        anotations,
        setAnotations: setAnotationsFn,
        getStorage,
        setStorage: saveToStorage,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  )
}

export function useStorage() {
  return useContext(LocalStorageContext)
}
