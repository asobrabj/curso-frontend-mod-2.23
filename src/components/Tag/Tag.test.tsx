import { fireEvent, waitFor } from '@testing-library/react'
import Tag from '.'
import { useStorage } from '../../context/localStorage/localStorage'
import { renderTheme } from '../../utils/styles-test'

// Mock do hook do contexto
jest.mock('../../context/localStorage/localStorage', () => ({
  ...jest.requireActual('../../context/localStorage/localStorage'), // Preserva as implementações não-mockadas
  useStorage: () => ({
    getTags: jest.fn(),
  }),
}))

describe('Tag Component', () => {
  test('Renderiza corretamente', () => {
    const { getByText } = renderTheme(
      <Tag name="categoria" text="Categoria 1" background="#ff0000" />
    )

    expect(getByText('Categoria 1')).toBeInTheDocument()
  })

  test('Chama a função getTags ao ser clicado', () => {
    const { getByText } = renderTheme(
      <Tag name="categoria" text="Categoria 1" background="#ff0000" />
    )
    const { getTags } = useStorage()

    fireEvent.click(getByText('Categoria 1'))
    waitFor(() => {
      expect(getTags).toHaveBeenCalledWith('category', 'Categoria 1')
    })
  })

  it('snapshot', () => {
    const { container } = renderTheme(<Tag />)

    expect(container).toMatchSnapshot()
  })
})
