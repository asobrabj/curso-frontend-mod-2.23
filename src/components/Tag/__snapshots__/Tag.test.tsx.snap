// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Component snapshot 1`] = `
.c0 {
  text-transform: capitalize;
  padding: 2px 9px;
  border: 1px solid #999;
  border-radius: 10px;
  font-weight: bold;
  margin-right: 8px;
  cursor: pointer;
}

<div>
   
  <span
    class="c0"
  />
   
</div>
`;
